src/core/labeling/rules/qgslabelingenginerule.h

```diff
@@ -0,0 +1,299 @@
+     *
+     * The returned string may contain HTML formatting.
+     */
+    virtual QString description() const;
+
+    /**
+     * Returns the name for this instance of the rule.
+     *
+     * The name is a user-configurable value which helps them identify and describe the
+     * rule within their projects.
+     *
+     * \see setName()
+     */
+    QString name() const { return mName; }
+
+    /**
+     * Sets the \a name for this instance of the rule.
+     *
+     * The name is a user-configurable value which helps them identify and describe the
+     * rule within their projects.
+     *
+     * \see name()
+     */
+    void setName( const QString &name ) { mName = name; }
+
+    /**
+     * Returns TRUE if the rule is active.
+     *
+     * \see setActive()
+     */
+    bool active() const;
+
+    /**
+     * Sets whether the rule is \a active.
+     *
+     * \see active()
+     */
+    void setActive( bool active );
+
+    /**
+     * Prepares the rule.
+     *
+     * This must be called on the main render thread, prior to commencing the render operation. Thread sensitive
+     * logic (such as creation of feature sources) can be performed in this method.
+     */
+    virtual bool prepare( QgsRenderContext &context ) = 0;
+
+    /**
+     * Writes the rule properties to an XML \a element.
+     *
+     * \see readXml()
+     */
+    virtual void writeXml( QDomDocument &doc, QDomElement &element, const QgsReadWriteContext &context ) const = 0;
+
+    /**
+     * Reads the rule properties from an XML \a element.
+     *
+     * \see resolveReferences()
+     * \see writeXml()
+     */
+    virtual void readXml( const QDomElement &element, const QgsReadWriteContext &context ) = 0;
+
+    /**
+     * Resolves reference to layers from stored layer ID.
+     *
+     * Should be called following a call readXml().
+     */
+    virtual void resolveReferences( const QgsProject *project );
+
+    /**
+     * Returns TRUE if a labeling candidate \a lp1 conflicts with \a lp2 after applying the rule.
+     *
+     * The default implementation returns FALSE.
+     */
+    virtual bool candidatesAreConflicting( const pal::LabelPosition *lp1, const pal::LabelPosition *lp2 ) const SIP_SKIP;
+
+    /**
+     * Returns a (possibly expanded) bounding box to use when searching for conflicts for a candidate.
+     *
+     * The return value is permitted to grow the bounding box, but may NOT shrink it.
+     *
+     * The default implementation returns the same bounds.
+     */
+    virtual QgsRectangle modifyCandidateConflictSearchBoundingBox( const QgsRectangle &candidateBounds ) const SIP_SKIP;
+
+    /**
+     * Returns TRUE if a labeling \a candidate violates the rule and should be eliminated.
+     *
+     * The default implementation returns FALSE.
+     */
+    virtual bool candidateIsIllegal( const pal::LabelPosition *candidate, QgsLabelingEngineContext &context ) const SIP_SKIP;
```

Comment:
I'm not sure why but I don't like the word 'illegal' here. What about candidateIsForbidden, candiateIsProhibited, candidateIsDisallowed, candidateIsBanned ? We're breaching rules not laws :wink:

Reply:
Hmm... we have `QgsAbstractDatabaseProviderConnection::illegalFieldNames`, but nothing in current api with forbidden/prohibited/disallowed/banned. Is one other method enough to set the precendent?

Reply:
@nyalldawson , ideally we'd change illegalFieldNames to prohibited/forbiddenFileNames but that API is exposed to bindings so it's frozen right? 

Seems like settling for what preexisting keyword is a good argument enough then.